File: src/mathproofmesh/communication/route_registry.py
from __future__ import annotations

import re
from typing import Any, Iterable

from ..config import SystemConfig
from ..schemas import (
    RouteDescriptor,
    RouteMember,
    RouteRole,
    RouteStatus,
    StrategyCard,
    stable_hash,
)


class RouteRegistry:
    """Authoritative route identity, role membership, and sparse-neighbor registry."""

    def __init__(
        self,
        config: SystemConfig | None = None,
        *,
        problem_hash: str = "",
        routes: Iterable[RouteDescriptor] = (),
    ) -> None:
        self.config = config
        self.problem_hash = problem_hash
        self._routes = {route.route_id: route for route in routes}
        self._strategy_index = {
            route.strategy_id: route.route_id for route in self._routes.values()
        }
        self._strategy_aliases: dict[str, str] = {}
        self.cooling_reasons: dict[str, str] = {}

    @property
    def routes(self) -> list[RouteDescriptor]:
        return list(self._routes.values())

    def get(self, route_id: str) -> RouteDescriptor:
        try:
            return self._routes[route_id]
        except KeyError as exc:
            raise KeyError(f"unknown route: {route_id}") from exc

    def register_route(
        self,
        strategy: StrategyCard,
        *,
        route_id: str | None = None,
    ) -> RouteDescriptor:
        existing_route_id = self._strategy_index.get(strategy.strategy_id)
        if existing_route_id is not None:
            return self.get(existing_route_id)
        # An explicit route_id is authoritative during checkpoint migration and
        # deterministic component fixtures. Ordinary production registration
        # omits it and therefore receives semantic de-duplication.
        duplicate = self.find_semantic_duplicate(strategy) if route_id is None else None
        if duplicate is not None:
            self._strategy_aliases[strategy.strategy_id] = duplicate.route_id
            duplicate.mechanism_signature = list(
                dict.fromkeys(
                    [
                        *duplicate.mechanism_signature,
                        *strategy.tags,
                        strategy.title,
                        *strategy.expected_lemmas,
                    ]
                )
            )
            return duplicate
        identifier = route_id or (
            "route_" + stable_hash((self.problem_hash, strategy.strategy_id))[:20]
        )
        if identifier in self._routes:
            existing = self._routes[identifier]
            if existing.strategy_id != strategy.strategy_id:
                raise ValueError("route ID already belongs to another strategy")
            return existing
        mechanism = list(
            dict.fromkeys([*strategy.tags, strategy.title, *strategy.expected_lemmas])
        )
        descriptor = RouteDescriptor(
            route_id=identifier,
            strategy_id=strategy.strategy_id,
            mechanism_signature=mechanism,
            strategy_signature=self.strategy_signature(strategy),
            shared_assumptions=list(dict.fromkeys(strategy.prerequisites)),
            inspiration_proposal_id=strategy.inspiration_proposal_id,
        )
        self._routes[identifier] = descriptor
        self._strategy_index[strategy.strategy_id] = identifier
        self.recompute_neighbors()
        return descriptor

    @staticmethod
    def _tokens(text: str) -> set[str]:
        return set(re.findall(r"[a-z0-9_]+|[\u4e00-\u9fff]", text.casefold()))

    @classmethod
    def strategy_signature(cls, strategy: StrategyCard) -> str:
        return stable_hash(
            {
                "title": " ".join(sorted(cls._tokens(strategy.title))),
                "core": " ".join(sorted(cls._tokens(strategy.core_idea))),
                "tags": sorted(set(item.casefold() for item in strategy.tags)),
                "prerequisites": sorted(
                    set(item.casefold() for item in strategy.prerequisites)
                ),
            }
        )

    def find_semantic_duplicate(self, strategy: StrategyCard) -> RouteDescriptor | None:
        signature = self.strategy_signature(strategy)
        exact = next(
            (
                route
                for route in self._routes.values()
                if route.strategy_signature and route.strategy_signature == signature
            ),
            None,
        )
        if exact is not None:
            return exact
        threshold = (
            self.config.topology.broker.duplicate_strategy_threshold
            if self.config is not None
            else 0.90
        )
        candidate_tokens = self._tokens(
            " ".join(
                [
                    strategy.title,
                    strategy.core_idea,
                    strategy.falsification_test,
                    *strategy.tags,
                    *strategy.prerequisites,
                ]
            )
        )
        if not candidate_tokens:
            return None
        best: tuple[float, RouteDescriptor] | None = None
        for route in self._routes.values():
            if route.status == RouteStatus.MERGED:
                continue
            route_tokens = self._tokens(" ".join(route.mechanism_signature))
            score = len(candidate_tokens & route_tokens) / max(
                1, len(candidate_tokens | route_tokens)
            )
            if best is None or score > best[0]:
                best = (score, route)
        return best[1] if best is not None and best[0] >= threshold else None

    def route_for_strategy(self, strategy_id: str) -> RouteDescriptor | None:
        route_id = self._strategy_index.get(strategy_id) or self._strategy_aliases.get(
            strategy_id
        )
        return self._routes.get(route_id) if route_id is not None else None

    def assign_member(
        self,
        route_id: str,
        agent_id: str,
        role: RouteRole,
        round_index: int,
    ) -> None:
        route = self.get(route_id)
        for member in route.members:
            if member.agent_id == agent_id and member.role == role:
                return
        maximum = (
            self.config.topology.route_teams.max_members_per_route
            if self.config is not None
            else 8
        )
        unique_agents = {member.agent_id for member in route.members}
        if agent_id not in unique_agents and len(unique_agents) >= maximum:
            raise ValueError(f"route {route_id} has reached its member limit")
        route.members.append(
            RouteMember(agent_id=agent_id, role=role, assigned_round=round_index)
        )

    def owns_agent(
        self, route_id: str, agent_id: str, role: RouteRole | None = None
    ) -> bool:
        route = self._routes.get(route_id)
        if route is None:
            return False
        return any(
            member.agent_id == agent_id and (role is None or member.role == role)
            for member in route.members
        )

    def set_neighbors(self, route_id: str, neighbors: list[str]) -> None:
        route = self.get(route_id)
        unique = [
            item
            for item in dict.fromkeys(neighbors)
            if item != route_id and item in self._routes
        ]
        limit = (
            self.config.topology.cross_route.max_neighbors_per_route
            if self.config is not None
            else len(unique)
        )
        route.neighbor_route_ids = unique[:limit]

    def recompute_neighbors(self) -> None:
        active = [
            route
            for route in self._routes.values()
            if route.status == RouteStatus.ACTIVE
        ]
        limit = (
            self.config.topology.cross_route.max_neighbors_per_route
            if self.config is not None
            else max(0, len(active) - 1)
        )
        for route in self._routes.values():
            if route.status != RouteStatus.ACTIVE:
                route.neighbor_route_ids = []
                continue
            source = self._tokens(" ".join(route.mechanism_signature))
            scored: list[tuple[float, str]] = []
            for candidate in active:
                if candidate.route_id == route.route_id:
                    continue
                target = self._tokens(" ".join(candidate.mechanism_signature))
                similarity = len(source & target) / max(1, len(source | target))
                # Sparse exchange benefits from some overlap, while deterministic
                # route_id tie-breaking preserves reproducibility.
                scored.append((similarity, candidate.route_id))
            scored.sort(key=lambda item: (-item[0], item[1]))
            route.neighbor_route_ids = [item[1] for item in scored[:limit]]

    def neighbors(self, route_id: str, round_index: int | None = None) -> list[str]:
        route = self.get(route_id)
        active = {
            item.route_id
            for item in self.active_routes(round_index or 0)
            if item.route_id != route_id
        }
        return [item for item in route.neighbor_route_ids if item in active]

    def mark_cooling(
        self,
        route_id: str,
        until_round: int,
        reason: str,
        *,
        requires_revision: bool = False,
    ) -> None:
        route = self.get(route_id)
        route.status = RouteStatus.COOLING
        route.cooldown_until_round = until_round
        route.requires_revision = requires_revision
        route.revision_summary = None
        self.cooling_reasons[route_id] = reason
        self.recompute_neighbors()

    def reactivate(self, route_id: str, *, revision_summary: str) -> None:
        route = self.get(route_id)
        if route.requires_revision and not revision_summary.strip():
            raise ValueError(
                "a counterexample-dependent route requires an explicit revision"
            )
        route.status = RouteStatus.ACTIVE
        route.cooldown_until_round = None
        route.requires_revision = False
        route.revision_summary = revision_summary.strip()
        self.recompute_neighbors()

    def merge_routes(self, source_route_id: str, target_route_id: str) -> None:
        if source_route_id == target_route_id:
            raise ValueError("a route cannot be merged into itself")
        source = self.get(source_route_id)
        target = self.get(target_route_id)
        source.status = RouteStatus.MERGED
        source.merged_into_route_id = target_route_id
        target.mechanism_signature = list(
            dict.fromkeys(target.mechanism_signature + source.mechanism_signature)
        )
        for strategy_id, route_id in list(self._strategy_aliases.items()):
            if route_id == source_route_id:
                self._strategy_aliases[strategy_id] = target_route_id
        self.recompute_neighbors()

    def active_routes(self, round_index: int) -> list[RouteDescriptor]:
        for route in self._routes.values():
            if (
                route.status == RouteStatus.COOLING
                and route.cooldown_until_round is not None
                and route.cooldown_until_round <= round_index
                and not route.requires_revision
            ):
                route.status = RouteStatus.ACTIVE
                route.cooldown_until_round = None
        self.recompute_neighbors()
        return [
            route
            for route in self._routes.values()
            if route.status == RouteStatus.ACTIVE
        ]

    def export_state(self) -> dict[str, Any]:
        return {
            "problem_hash": self.problem_hash,
            "routes": [route.model_dump(mode="json") for route in self.routes],
            "cooling_reasons": dict(self.cooling_reasons),
            "strategy_aliases": dict(self._strategy_aliases),
        }

    @classmethod
    def from_state(
        cls,
        state: dict[str, Any],
        config: SystemConfig | None = None,
    ) -> "RouteRegistry":
        registry = cls(
            config,
            problem_hash=str(state.get("problem_hash", "")),
            routes=[
                RouteDescriptor.model_validate(item) for item in state.get("routes", [])
            ],
        )
        registry.cooling_reasons = {
            str(key): str(value)
            for key, value in dict(state.get("cooling_reasons", {})).items()
        }
        registry._strategy_aliases = {
            str(key): str(value)
            for key, value in dict(state.get("strategy_aliases", {})).items()
        }
        return registry

File: mathproofmesh-core/src/main/java/io/github/aililuola/mathproofmesh/communication/RouteRegistry.java
package io.github.aililuola.mathproofmesh.communication;

import io.github.aililuola.mathproofmesh.contract.CanonicalJson;
import io.github.aililuola.mathproofmesh.contract.RouteDescriptor;
import io.github.aililuola.mathproofmesh.contract.RouteMember;
import io.github.aililuola.mathproofmesh.contract.RouteRole;
import io.github.aililuola.mathproofmesh.contract.RouteStatus;
import io.github.aililuola.mathproofmesh.contract.StrategyCard;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class RouteRegistry {
  private static final Pattern TOKEN_PATTERN =
      Pattern.compile("[a-z0-9_]+|[\\u4e00-\\u9fff]");
  private static final Set<RouteStatus> SCHEDULABLE =
      Set.of(RouteStatus.ACTIVE, RouteStatus.REPAIR_ONCE);

  private final String problemHash;
  private final int maxNeighbors;
  private final int maxMembers;
  private final double duplicateThreshold;
  private final Map<String, RouteDescriptor> routes = new LinkedHashMap<>();
  private final Map<String, String> strategyIndex = new LinkedHashMap<>();
  private final Map<String, String> strategyAliases = new LinkedHashMap<>();

  public RouteRegistry(
      String problemHash, int maxNeighbors, int maxMembers, double duplicateThreshold) {
    if (problemHash == null || problemHash.isBlank()) {
      throw new IllegalArgumentException("problemHash is required");
    }
    if (maxNeighbors < 0 || maxMembers < 1) {
      throw new IllegalArgumentException("route limits are invalid");
    }
    if (duplicateThreshold < 0.0 || duplicateThreshold > 1.0) {
      throw new IllegalArgumentException("duplicateThreshold must be between zero and one");
    }
    this.problemHash = problemHash.strip();
    this.maxNeighbors = maxNeighbors;
    this.maxMembers = maxMembers;
    this.duplicateThreshold = duplicateThreshold;
  }

  public String problemHash() {
    return problemHash;
  }

  public synchronized List<RouteDescriptor> routes() {
    return List.copyOf(routes.values());
  }

  public synchronized RouteDescriptor get(String routeId) {
    RouteDescriptor route = routes.get(routeId);
    if (route == null) {
      throw new IllegalArgumentException("unknown route: " + routeId);
    }
    return route;
  }

  public synchronized boolean exists(String routeId) {
    return routes.containsKey(routeId);
  }

  public synchronized RouteDescriptor register(RouteDescriptor descriptor) {
    Objects.requireNonNull(descriptor, "descriptor");
    RouteDescriptor existing = routes.get(descriptor.routeId());
    if (existing != null) {
      if (!existing.strategyId().equals(descriptor.strategyId())) {
        throw new IllegalArgumentException("route ID already belongs to another strategy");
      }
      return existing;
    }
    String indexed = strategyIndex.get(descriptor.strategyId());
    if (indexed != null) {
      return routes.get(indexed);
    }
    routes.put(descriptor.routeId(), descriptor);
    strategyIndex.put(descriptor.strategyId(), descriptor.routeId());
    recomputeNeighbors();
    return routes.get(descriptor.routeId());
  }

  public synchronized RouteDescriptor registerStrategy(StrategyCard strategy) {
    Objects.requireNonNull(strategy, "strategy");
    String indexed = strategyIndex.get(strategy.strategyId());
    if (indexed != null) {
      return routes.get(indexed);
    }
    Optional<RouteDescriptor> duplicate = findSemanticDuplicate(strategy);
    if (duplicate.isPresent()) {
      strategyAliases.put(strategy.strategyId(), duplicate.orElseThrow().routeId());
      return duplicate.orElseThrow();
    }
    String routeId =
        "route_"
            + CanonicalJson.stableHash(List.of(problemHash, strategy.strategyId()))
                .substring(0, 20);
    List<String> mechanism = new ArrayList<>();
    mechanism.addAll(strategy.tags());
    mechanism.add(strategy.title());
    mechanism.addAll(strategy.expectedLemmas());
    RouteDescriptor descriptor =
        new RouteDescriptor(
            null,
            0,
            0,
            null,
            null,
            strategy.inspirationProposalId(),
            null,
            null,
            null,
            distinct(mechanism),
            List.of(),
            null,
            List.of(),
            0,
            false,
            null,
            routeId,
            distinct(strategy.prerequisites()),
            0,
            RouteStatus.ACTIVE,
            strategy.strategyId(),
            strategySignature(strategy));
    return register(descriptor);
  }

  public synchronized Optional<RouteDescriptor> routeForStrategy(String strategyId) {
    String routeId = strategyIndex.get(strategyId);
    if (routeId == null) {
      routeId = strategyAliases.get(strategyId);
    }
    return Optional.ofNullable(routes.get(routeId));
  }

  public synchronized void assignMember(
      String routeId, String agentId, RouteRole role, int roundIndex) {
    RouteDescriptor route = get(routeId);
    if (route.members().stream()
        .anyMatch(member -> member.agentId().equals(agentId) && member.role() == role)) {
      return;
    }
    long uniqueMembers =
        route.members().stream().map(RouteMember::agentId).distinct().count();
    boolean alreadyAssigned =
        route.members().stream().anyMatch(member -> member.agentId().equals(agentId));
    if (!alreadyAssigned && uniqueMembers >= maxMembers) {
      throw new IllegalStateException("route " + routeId + " has reached its member limit");
    }
    List<RouteMember> members = new ArrayList<>(route.members());
    members.add(new RouteMember(agentId, roundIndex, role));
    routes.put(routeId, copy(route, List.copyOf(members), route.neighborRouteIds(), route.status()));
  }

  public synchronized boolean ownsAgent(String routeId, String agentId, RouteRole role) {
    RouteDescriptor route = routes.get(routeId);
    return route != null
        && route.members().stream()
            .anyMatch(
                member ->
                    member.agentId().equals(agentId)
                        && (role == null || member.role() == role));
  }

  public synchronized void setNeighbors(String routeId, Collection<String> candidates) {
    RouteDescriptor route = get(routeId);
    List<String> selected =
        candidates.stream()
            .filter(candidate -> !candidate.equals(routeId))
            .filter(routes::containsKey)
            .distinct()
            .limit(maxNeighbors)
            .toList();
    routes.put(routeId, copy(route, route.members(), selected, route.status()));
  }

  public synchronized List<String> neighbors(String routeId) {
    RouteDescriptor route = get(routeId);
    return route.neighborRouteIds().stream()
        .filter(routes::containsKey)
        .filter(candidate -> SCHEDULABLE.contains(routes.get(candidate).status()))
        .toList();
  }

  public synchronized void markCooling(String routeId, int untilRound, boolean requiresRevision) {
    RouteDescriptor route = get(routeId);
    routes.put(
        routeId,
        new RouteDescriptor(
            untilRound,
            route.duplicateAttemptCount(),
            route.failureCount(),
            route.frozenReason(),
            route.frozenSignature(),
            route.inspirationProposalId(),
            route.lastProgressSignature(),
            route.latestAttemptId(),
            route.latestCheckpointId(),
            route.mechanismSignature(),
            route.members(),
            route.mergedIntoRouteId(),
            route.neighborRouteIds(),
            route.noProgressStrikes(),
            requiresRevision,
            null,
            route.routeId(),
            route.sharedAssumptions(),
            route.stagnationRounds(),
            RouteStatus.COOLING,
            route.strategyId(),
            route.strategySignature()));
    recomputeNeighbors();
  }

  public synchronized void reactivate(String routeId, String revisionSummary) {
    RouteDescriptor route = get(routeId);
    if (route.requiresRevision() && (revisionSummary == null || revisionSummary.isBlank())) {
      throw new IllegalArgumentException(
          "a counterexample-dependent route requires an explicit revision");
    }
    routes.put(
        routeId,
        new RouteDescriptor(
            null,
            route.duplicateAttemptCount(),
            route.failureCount(),
            null,
            null,
            route.inspirationProposalId(),
            route.lastProgressSignature(),
            route.latestAttemptId(),
            route.latestCheckpointId(),
            route.mechanismSignature(),
            route.members(),
            route.mergedIntoRouteId(),
            route.neighborRouteIds(),
            0,
            false,
            revisionSummary == null ? "" : revisionSummary.strip(),
            route.routeId(),
            route.sharedAssumptions(),
            route.stagnationRounds(),
            RouteStatus.ACTIVE,
            route.strategyId(),
            route.strategySignature()));
    recomputeNeighbors();
  }

  public synchronized void merge(String sourceRouteId, String targetRouteId) {
    if (sourceRouteId.equals(targetRouteId)) {
      throw new IllegalArgumentException("a route cannot be merged into itself");
    }
    RouteDescriptor source = get(sourceRouteId);
    RouteDescriptor target = get(targetRouteId);
    routes.put(
        sourceRouteId,
        new RouteDescriptor(
            source.cooldownUntilRound(),
            source.duplicateAttemptCount(),
            source.failureCount(),
            source.frozenReason(),
            source.frozenSignature(),
            source.inspirationProposalId(),
            source.lastProgressSignature(),
            source.latestAttemptId(),
            source.latestCheckpointId(),
            source.mechanismSignature(),
            source.members(),
            targetRouteId,
            List.of(),
            source.noProgressStrikes(),
            source.requiresRevision(),
            source.revisionSummary(),
            source.routeId(),
            source.sharedAssumptions(),
            source.stagnationRounds(),
            RouteStatus.MERGED,
            source.strategyId(),
            source.strategySignature()));
    List<String> combined = new ArrayList<>(target.mechanismSignature());
    combined.addAll(source.mechanismSignature());
    routes.put(
        targetRouteId,
        new RouteDescriptor(
            target.cooldownUntilRound(),
            target.duplicateAttemptCount(),
            target.failureCount(),
            target.frozenReason(),
            target.frozenSignature(),
            target.inspirationProposalId(),
            target.lastProgressSignature(),
            target.latestAttemptId(),
            target.latestCheckpointId(),
            distinct(combined),
            target.members(),
            target.mergedIntoRouteId(),
            target.neighborRouteIds(),
            target.noProgressStrikes(),
            target.requiresRevision(),
            target.revisionSummary(),
            target.routeId(),
            target.sharedAssumptions(),
            target.stagnationRounds(),
            target.status(),
            target.strategyId(),
            target.strategySignature()));
    recomputeNeighbors();
  }

  public synchronized void activateCooledRoutes(int currentRound) {
    for (RouteDescriptor route : List.copyOf(routes.values())) {
      if (route.status() == RouteStatus.COOLING
          && route.cooldownUntilRound() != null
          && route.cooldownUntilRound() <= currentRound
          && !route.requiresRevision()) {
        reactivate(route.routeId(), "");
      }
    }
  }

  public synchronized void recomputeNeighbors() {
    List<RouteDescriptor> active =
        routes.values().stream().filter(route -> SCHEDULABLE.contains(route.status())).toList();
    for (RouteDescriptor route : List.copyOf(routes.values())) {
      if (!SCHEDULABLE.contains(route.status())) {
        routes.put(route.routeId(), copy(route, route.members(), List.of(), route.status()));
        continue;
      }
      Set<String> sourceTokens = tokens(String.join(" ", route.mechanismSignature()));
      List<String> selected =
          active.stream()
              .filter(candidate -> !candidate.routeId().equals(route.routeId()))
              .sorted(
                  Comparator.<RouteDescriptor>comparingDouble(
                          candidate ->
                              -similarity(
                                  sourceTokens,
                                  tokens(String.join(" ", candidate.mechanismSignature()))))
                      .thenComparing(RouteDescriptor::routeId))
              .limit(maxNeighbors)
              .map(RouteDescriptor::routeId)
              .toList();
      routes.put(route.routeId(), copy(route, route.members(), selected, route.status()));
    }
  }

  private Optional<RouteDescriptor> findSemanticDuplicate(StrategyCard strategy) {
    String signature = strategySignature(strategy);
    Optional<RouteDescriptor> exact =
        routes.values().stream()
            .filter(route -> route.strategySignature().equals(signature))
            .findFirst();
    if (exact.isPresent()) {
      return exact;
    }
    Set<String> candidateTokens =
        tokens(
            String.join(
                " ",
                strategy.title(),
                strategy.coreIdea(),
                strategy.falsificationTest(),
                String.join(" ", strategy.tags()),
                String.join(" ", strategy.prerequisites())));
    return routes.values().stream()
        .filter(route -> route.status() != RouteStatus.MERGED)
        .map(
            route ->
                Map.entry(
                    route,
                    similarity(
                        candidateTokens, tokens(String.join(" ", route.mechanismSignature())))))
        .filter(entry -> entry.getValue() >= duplicateThreshold)
        .max(
            Comparator.<Map.Entry<RouteDescriptor, Double>>comparingDouble(Map.Entry::getValue)
                .thenComparing(entry -> entry.getKey().routeId()))
        .map(Map.Entry::getKey);
  }

  private static String strategySignature(StrategyCard strategy) {
    Map<String, Object> payload = new LinkedHashMap<>();
    payload.put("title", sortedTokenText(strategy.title()));
    payload.put("core", sortedTokenText(strategy.coreIdea()));
    payload.put(
        "tags",
        strategy.tags().stream().map(value -> value.toLowerCase(Locale.ROOT)).distinct().sorted().toList());
    payload.put(
        "prerequisites",
        strategy.prerequisites().stream()
            .map(value -> value.toLowerCase(Locale.ROOT))
            .distinct()
            .sorted()
            .toList());
    return CanonicalJson.stableHash(payload);
  }

  private static String sortedTokenText(String value) {
    return String.join(" ", tokens(value).stream().sorted().toList());
  }

  private static Set<String> tokens(String value) {
    Matcher matcher = TOKEN_PATTERN.matcher(value.toLowerCase(Locale.ROOT));
    Set<String> result = new LinkedHashSet<>();
    while (matcher.find()) {
      result.add(matcher.group());
    }
    return result;
  }

  private static double similarity(Set<String> left, Set<String> right) {
    if (left.isEmpty() && right.isEmpty()) {
      return 1.0;
    }
    Set<String> union = new LinkedHashSet<>(left);
    union.addAll(right);
    Set<String> intersection = new LinkedHashSet<>(left);
    intersection.retainAll(right);
    return (double) intersection.size() / Math.max(1, union.size());
  }

  private static List<String> distinct(Collection<String> values) {
    return List.copyOf(new LinkedHashSet<>(values));
  }

  private static RouteDescriptor copy(
      RouteDescriptor route,
      List<RouteMember> members,
      List<String> neighbors,
      RouteStatus status) {
    return new RouteDescriptor(
        route.cooldownUntilRound(),
        route.duplicateAttemptCount(),
        route.failureCount(),
        route.frozenReason(),
        route.frozenSignature(),
        route.inspirationProposalId(),
        route.lastProgressSignature(),
        route.latestAttemptId(),
        route.latestCheckpointId(),
        route.mechanismSignature(),
        members,
        route.mergedIntoRouteId(),
        neighbors,
        route.noProgressStrikes(),
        route.requiresRevision(),
        route.revisionSummary(),
        route.routeId(),
        route.sharedAssumptions(),
        route.stagnationRounds(),
        status,
        route.strategyId(),
        route.strategySignature());
  }
}
